scripts: build meta file workspace status extension

tejlmand opened the pull request:
The PR #39382 raised a discussion on build reproducibility and knowledge
of west projects being out of sync with the west manifest.

Similar to how `git submodules` will report the working tree dirty if
any of the submodules HEAD points to a SHA different than the one
recorded in the super project.

Based on this discussion this commit extends the Zephyr build meta file
with overall workspace status included in the meta file.

It add the following meta info to the build meta file:
> workspace:
>  dirty: false / true
>  extra: false / true
>  off: false / true

A project using west and having an extra Zephyr module loaded not
controlled using git and a west project at a SHA different than the
SHA referenced by the manifest can look like:
```
zephyr:
  path: /.../zephyr
  revision: 863600c
modules:
- name: mcuboot
  path: /.../bootloader/mcuboot
  revision: c61538748ead773ea75a551a7beee299228bdcaf
- name: local_module
  path: /.../local_module
  revision: null
west:
  manifest: /.../zephyr/west.yml
  projects:
  - path: /.../zephyr
    revision: 863600c
  - path: /.../bootloader/mcuboot
    revision: c61538748ead773ea75a551a7beee299228bdcaf-off
  - path: /.../tools/net-tools
    revision: f49bd1354616fae4093bf36e5eaee43c51a55127
workspace:
  dirty: false
  extra: true
  'off': true
```

And without west:
```
zephyr:
  path: /.../zephyr
  revision: 863600c
modules:
- name: hal_nordic
  path: /.../modules/hal/nordic
  revision: a6e5299041f152da5ae0ab17b2e44e088bb96d6d
- name: local_module
  path: /.../local_module
  revision: null
west: null
workspace:
  dirty: false
  extra: true
  'off': false
```

The workspace status can be propagated to the Zephyr revision, where the final revision will become: <SHA>-<state1>-<state2>-<state3>

For example:
> zephyr:
>  path: /.../zephyr
>  revision: <SHA>-dirty-extra-off

or
> zephyr:
>  path: /.../zephyr
>  revision: <SHA>-extra

The `BUILD_OUTPUT_META_STATE_PROPAGATE` Kconfig setting is introduced
to provide user control of this behavior.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand committed:
scripts: build meta file workspace status extension

The PR #39382 raised a discussion on build reproducibility and knowledge
of west projects being out of sync with the west manifest.

Similar to how `git submodules` will report the working tree dirty if
any of the submodules HEAD points to a SHA different than the one
recorded in the super project.

Based on this discussion this commit extends the Zephyr build meta file
with overall workspace status included in the meta file.

It adds the following meta info to the build meta file:
> workspace:
>  dirty: false / true
>  extra: false / true
>  off:   false / true

A project using west and having an extra Zephyr module loaded not
controlled using git and a west project at a SHA different than the
SHA referenced by the manifest can look like:
zephyr:
  path: /.../zephyr
  revision: 863600cd0e3c0a271e86629c5089821e5e4380cc
modules:
- name: mcuboot
  path: /.../bootloader/mcuboot
  revision: c61538748ead773ea75a551a7beee299228bdcaf
- name: local_module
  path: /.../local_module
  revision: null
west:
  manifest: /.../zephyr/west.yml
  projects:
  - path: /.../zephyr
    revision: 863600cd0e3c0a271e86629c5089821e5e4380cc
  - path: /.../bootloader/mcuboot
    revision: c61538748ead773ea75a551a7beee299228bdcaf-off
  - path: /.../tools/net-tools
    revision: f49bd1354616fae4093bf36e5eaee43c51a55127
workspace:
  dirty: false
  extra: true
  'off': true

And without west:
zephyr:
  path: /.../zephyr
  revision: 863600cd0e3c0a271e86629c5089821e5e4380cc
modules:
- name: hal_nordic
  path: /.../modules/hal/nordic
  revision: a6e5299041f152da5ae0ab17b2e44e088bb96d6d
- name: local_module
  path: /.../local_module
  revision: null
workspace:
  dirty: false
  extra: true

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

SebastianBoe commented:
There is a lot about what we are doing, but not so much about why.

Tracking backwards I eventually get to:

"In order to ensure that a Zephyr application is always built from the exact same sources it would be good if west could report if any of the west-handled repositories is out of sync with the manifest file. Much like a git superproject git status would report a dirty working tree if any of the submodules' HEADs point to a sha different from the one recorded in the superproject."

which is asking for a west equivalent of "git status". git status does not produce any files, so I am lacking the motivation here. I don't want to read a long discussion. The why belongs in the commit message IMHO.

marc-hb commented:
> Tracking backwards I eventually get to: [...] which is asking for a west equivalent of "git status".

Tracking every more backwards, the initial request for a `west status` (or other) command similar to `git status`: https://github.com/zephyrproject-rtos/west/issues/548

> git status does not produce any files

Well maybe it should because its output for submodules is not easy to parse, is it? Digression sorry.

tejlmand committed:
scripts: support propagating workspace status to zephyr revision

The PR #39382 raised a discussion on build reproducibility and knowledge
of west projects being out of sync with the west manifest.

Similar to how `git submodules` will report the working tree dirty if
any of the submodules HEAD points to a SHA different than the one
recorded in the super project.

Based on this discussion this commit extends the Zephyr and manifest
repo (when `west` is used) revisions to include the concatenated states
of the workspace.

The workspace states are:
>  dirty: false / true
>  extra: false / true
>  off:   false / true

the final revision will become: <SHA>-<state1>-<state2>-<state3>

For example:
> zephyr:
>  path: /.../zephyr
>  revision: <SHA>-dirty-extra-off

or
> zephyr:
>  path: /.../zephyr
>  revision: <SHA>-extra

The `BUILD_OUTPUT_META_STATE_PROPAGATE` Kconfig setting is introduced
to provide user control of this behavior.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand commented:
> There is a lot about what we are doing, but not so much about why.

Commit message updated, thanks for the suggestion.

tejlmand commented:
> > Tracking backwards I eventually get to: [...] which is asking for a west equivalent of "git status".
> 
> Tracking every more backwards, the initial request for a `west status` (or other) command similar to `git status`: [zephyrproject-rtos/west#548](https://github.com/zephyrproject-rtos/west/issues/548)

which is referenced in this commit message:
https://github.com/zephyrproject-rtos/zephyr/pull/39382/commits/e5fd5910b402b1ee9e7e93a5f6ab7f22b2b78507
> This commit is motivated by the west discussion in:
zephyrproject-rtos/west#548
>
> The commit provides the abilty to generate a build meta info file
containing lists of:

marc-hb commented:
> which is referenced in this commit message: e5fd591

Sorry I missed that. Commit messages are 3rd class citizens on Github: it's tedious to look at them, it's not possible to review them, it's not possible to `range-diff` them at all,... The "backlinks" automatically created from commit messages  are also pretty useless, check the ones in west 548. [In "normal", non-Zephyr Github usage, commits are just fixups](https://github.com/isaacs/github/issues/959#issuecomment-592978474), no concept of a patch series.

I checked the description field and comments only because I expect descriptions to be either fairly complete duplicates of all commit messages or mostly empty and this description is very long and detailed (not sure why you omitted that link from the description). I'm more and more often leaving the description field in my PRs almost empty to force reviewers to look at the original commit messages even though it's less convenient. Focusing on git commit messages instead of Github descriptions also reduces "vendor lock-in" a bit.

Totally off-topic sorry (but IMHO important)

tejlmand commented:
> not sure why you omitted that link from the description while mentioning others

Because of this: https://github.com/zephyrproject-rtos/zephyr/pull/40085#issue-773338682
> continuation of: #39382
>
> Only second commit to be reviewed in this PR, first commit is under review here: #39382

The referred commit is the first commit in this PR and thus under review else-where.

> Commit messages are 3rd class citizens on Github: it's tedious to look at them, it's not possible to review them,

I also look at the commit messages, I find them useful when reviewing / understanding the code and important later if a blame or bisect is needed later.
But I agree github lacks a clean way of commenting the commit message, so I usually do this on the first line change in the commit: `nothing wrong on this line, but ...... something related to the commit message ......`

> [Mark pull request as depending on another](https://github.com/isaacs/github/issues/959#issuecomment-592978474)

Thanks for that link.
Would be very useful as long as they avoid the gerrit way of doing things.

mbolivar-nordic commented:
> Only second commit to be reviewed in this PR, first commit is under review here: #39382

there are 3 commits here; does this need rebasing?

tejlmand commented:
> > Only second commit to be reviewed in this PR, first commit is under review here: #39382
> 
> there are 3 commits here; does this need rebasing?

Yes, the dependent PR has been merged, so I have just rebased this one (which btw. now contains two commits).

mbolivar-nordic commented on Kconfig.zephyr:
```diff
@@ -512,7 +512,14 @@ config BUILD_OUTPUT_META
+	  - Workspace:
+	    - dirty: one or more repositories are marked dirty
+	    - extra: extra Zephyr modules was manually included in the build
+	    - off:   the SHA of one or more west projects does not match the
+	             referenced by the west manifest (west only state)
```
What does this mean? Does it mean that the HEAD commit is different than the manifest-rev commit? If so, I think we should just say that. If not, I think this needs to be made clearer; I can't understand what this is saying just from the Kconfig description.

tejlmand replied:
it is, because the head of the manifest-rev is what west returns as sha.
But are users of `west` familiar with the manifest-rev ?

What about:
```
- off:   the SHA of one or more west projects has diverged since the last `west update`
          and are therefore out of sync with the manifest (west only state).
```

mbolivar-nordic replied:
I think "west only state" is unnecessarily vague. That's why I said:

> Does it mean that the HEAD commit is different than the manifest-rev commit? If so, I think we should just say that.

I suggest:

```
- off:   the SHA of one or more west projects has diverged since the last `west update`
         (that is, the current commit is different than the manifest-rev commit)
```

tejlmand replied:
> I think "west only state" is unnecessarily vague. 

What that means is that this state is only present if `west` is used for workspace management.
I will not be present in non-west workspaces.

mbolivar-nordic replied:
>> I think "west only state" is unnecessarily vague.

> What that means is that this state is only present if west is used for workspace management.
> I will not be present in non-west workspaces.

Thanks for explaining, but that doesn't solve the problem with this pull request. My point is that this wording is vague and I have a suggestion for how to make it clearer. Writing a comment in the github pull request conversation clearing it up will not help people who are going to read the code later. What do you say to my suggestion?

tejlmand replied:
> My point is that this wording is vague and I have a suggestion for how to make it clearer.

Yes, I understand that.
And you're proposal is fine, but it lacks the part I also wanted to tell the reader, that this state can only be present if `west` is used, which were the point I tried to make with my last comment.

You may say that piece of information can be omitted, because it should be obvious when we refer to `west update` in the text, I just want to be sure you know the reason for that sentence so that it can be reworded, or we agree why it can be removed.

SebastianBoe replied:
(west only state) is not being explanatory unfortunately. It should either be omitted or re-written.

SebastianBoe replied:
same in the Kconfig help text

tejlmand replied:
> (west only state) is not being explanatory unfortunately. It should either be omitted or re-written.

I hope it's clear from earlier comments that I agree to this.
I just want to know whether omitting this or re-writing is preferred.

If there is consensus that this information is not relevant, then it's easier to remove.

SebastianBoe replied:
Whether it should be omitted or re-written depends on what you are trying to say.

You say that "the state is only present if west is used".

Does that mean that off : False, when west is not used, or that the attribute "off" is not present at all?

mbolivar-nordic replied:
I have the same question as @SebastianBoe .

tejlmand replied:
> Does that mean that off : False, when west is not used, or that the attribute "off" is not present at all?

Good point.
To make this clear, I have now ensured that `west` and statuses only related to `west` usage are now only present in the meta file when `west` is used.

So that means not present at all.

tejlmand replied:
Updated to consistent behavior, and description should be clearer now.

Thanks for the feedback.

carlescufi commented:
@SebastianBoe can you please review this one?

SebastianBoe reviewed (changes requested):
Some changes requested, the approach looks ok.

SebastianBoe commented on scripts/zephyr_module.py:
```diff
@@ -282,21 +287,41 @@ def git_revision(path):
+                workspace_off = True
             meta_project = {'path': project_path,
-                            'revision': git_revision(project_path)}
+                            'revision': revision}
```
The difference between the first iteration and the rest is not significant enought to merit the code duplication IMHO.

Let's iterate over an index of projects instead and add a if i == 0 for the special case.

tejlmand replied:
it's not only about code duplication, if you notice the project `0` is the manifest repo, and therefore results are stored in a dedicated variables. Those dedicated variables are used further down here:
https://github.com/zephyrproject-rtos/zephyr/blob/aa472a7476157b7ab830744058be62dbd2078fc7/scripts/zephyr_module.py#L351
https://github.com/zephyrproject-rtos/zephyr/blob/aa472a7476157b7ab830744058be62dbd2078fc7/scripts/zephyr_module.py#L352
https://github.com/zephyrproject-rtos/zephyr/blob/aa472a7476157b7ab830744058be62dbd2078fc7/scripts/zephyr_module.py#L357

Also, the code concerning the `MANIFEST_REV_BRANCH` is not to be called for the project `0` because it's the manifest repo and therefore doesn't have that particular branch, so you would endup with something like:
```
        for i, project in enumerate(projects):
            project_path = PurePath(project.posixpath).as_posix()
            revision, dirty = git_revision(project_path)
            workspace_dirty |= dirty
            if i > 0:
                # The manifest project doesn't have a MANIFEST_REV_BRANCH.
                if project.sha(MANIFEST_REV_BRANCH) != revision:
                    revision += '-off'
                    workspace_off = True

            meta_project = {'path': project_path,
                            'revision': revision}
            meta_projects.append(meta_project)
            if i == 0:
                manifest_revision = revision
                manifest_dirty = dirty
                manifest_project = meta_project
```

That is 15 LOC vs current 16 LOC, but you get an additional `enumerate(projects)` because you need to know when you are processing project `0`.

Most code lines saved are just changed into `manifest_<something> = <something>` instead of the clear `git_revision()` and `PurePath()` calls.

I don't really see the gain in changing this code, but will do so if more people prefers the alternative presented.
@mbolivar-nordic do you have any preference ?

SebastianBoe replied:
"therefore results are stored in a dedicated variables. Those dedicated variables are used further down here:"

I did not notice this, I assumed the comment was correct:

        # Special treatment of first project as it is the manifest project and
        # thus doesn't have a MANIFEST_REV_BRANCH.

Then I suggest that either the comment is reduced to "Special treatment of the first project", or, if you prefer, expanded to explain the full set of reasons that it must be treated specially.

tejlmand replied:
That comment is not wrong, but you're right that there is more to it than just this.
Will rephrase the comment.

tejlmand replied:
done

SebastianBoe commented on scripts/zephyr_module.py:
```diff
@@ -417,6 +464,10 @@ def main():
+    parser.add_argument('--meta-state-propagate', action='store_true',
+                        help="""Propagate state of modules and west projects
+                             to the Zephyr SHA and if west is used, to the
+                             manifest SHA""")
```
"to the suffix of the Zephyr SHA" maybe?

Propagating something to a SHA sounds weird.

tejlmand replied:
It sure does sound strange, thanks for spotting.

tejlmand replied:
done.

marc-hb commented:
The `-off` attribute (I like the name BTW)  implements https://github.com/zephyrproject-rtos/west/issues/548 exactly with two caveats:
1. It's available only for Zephyr and not for other west users
2. No report of which particular repos are "off"?

Two questions:
- Is 2. correct? I mean does this PR really not show which repos are "off"? I didn't see that in the commit message and examples but maybe I just missed it? If not, can this critical information be provided somehow?
- More generally speaking, could  the non-Zephyr specific features being added here be later moved outside Zephyr and made available to all west users? After some probation period? I believe @mbolivar-nordic mentioned this probation period idea once somewhere but in any case I'd like confirmation it applies here. Finally, are the non-Zephyr specific features here designed to avoid backward compatibility issues if/when they're transferred to west?

tejlmand commented:
> 1. It's available only for Zephyr and not for other west users

Correct, because it's located in Zephyr repo.
But the functionality itself is independent of Zephyr, so you can run it directly using:
```
$ <zephyr-repo>/scripts/zephyr_module.py --meta-out test.meta
```

which will produce a `test.meta` file with above status'es.
> More generally speaking, could the non-Zephyr specific features being added here be later moved outside Zephyr and made available to all west users? After some probation period?

It could. There is no special magic, and the `zephyr_module.py` script could then just ask `west` to append the status instead of doing it by itself.

mbolivar-nordic commented on Kconfig.zephyr:
```diff
@@ -512,7 +512,15 @@ config BUILD_OUTPUT_META
+	    - extra: extra Zephyr modules was manually included in the build
+	    - off:   the SHA of one or more west projects has diverged since the
+	             last `west update`.
+	             (`off` state is only present for west managed workspaces)
```
2 things:

1. Why are you using backticks in Kconfig help? This isn't markdown.

2. I don't think that Zephyr has a concept of a "workspace" apart from a "west workspace", at least not one that we have documented anywhere. Did I miss one? If not, then I think this is confusing wording and would suggest something like "the off state is only present if a west workspace is found"

tejlmand replied:
> Why are you using backticks in Kconfig help? This isn't markdown.

I know it's not mark down, but for the commands I wanted to indicate those are commands.
I considered both \`west update\` or 'west update' before I noticed we use \` a couple of places:
https://github.com/zephyrproject-rtos/zephyr/blob/fc2756cf1817e3ee579cd9abe3b3c579941a5752/drivers/serial/Kconfig.native_posix#L47
https://github.com/zephyrproject-rtos/zephyr/blob/fc2756cf1817e3ee579cd9abe3b3c579941a5752/drivers/ethernet/Kconfig.native_posix#L18

(although I notice now this Kconfig code uses \`\` and not a single \`)

but we also have '...':
https://github.com/zephyrproject-rtos/zephyr/blob/fc2756cf1817e3ee579cd9abe3b3c579941a5752/Kconfig.zephyr#L596
 
We don't seem to have a standard on this, so i'm fine with any proposal, \`...\`, \`\`...\`\`, '...', or something else.

@mbolivar-nordic do you have any preference ?

(I removed the \` on the \`off\` as this is not a command)

> I don't think that Zephyr has a concept of a "workspace" apart from a "west workspace",

I was more thinking that non-west users could have their own definition of a workspace, but `west workspace` is just as clear and precise, even for non-west users.

Fixed.

mbolivar-nordic reviewed (approved):
Fine with me, thanks. A couple of fixes suggested.

tejlmand commented:
> > More generally speaking, could the non-Zephyr specific features being added here be later moved outside Zephyr and made available to all west users? After some probation period?
> 
> It could. There is no special magic, and the `zephyr_module.py` script could then just ask `west` to append the status instead of doing it by itself.

@marc-hb one thing to consider before taking this directly to `west` is branches `revision: <branch>` and pull requests `revision: pull/<no>/head`.

As a build meta file, we are placing SHAs which is in order to uniquely reproduce a build.
Here it's sufficient to know the revision is not `off` as that could mean a locally available SHA only.
The fact that `west` tracks a branch is not a problem, as a valid SHA indicates the SHA could be somewhere on the branch.
(history rewrite on branches / PRs are users own responsibility)

But if this is integrated into `west` itself, there is a risk that people would expect that `west` can tell you if your branch in up-to-date (when tracking a branch).
That is not the purpose of this meta file. The purpose of the meta file is reproducibility.
(This is why the SHAs are important, including the knowledge whether extra Zephyr modules was added, dirty state, and if any of the `west` project are `off` == indication that SHA could be a local only SHA)

marc-hb commented:
> @marc-hb one thing to consider before taking this directly to west is branches revision: <branch> and pull requests revision: pull/<no>/head.

Thanks, branches (and worse complexities) were debated in zephyrproject-rtos/west#548  and the tl;dr was: comparing to `manifest-rev` is simple and not always but often useful (at least some people thought it is useful). Anything else is too complex and variable to be bug-free or meaningful even.

Comparing to `manifest-rev` seems to be what this PR does (no time for a detailed review sorry).

> But if this is integrated into west itself, there is a risk that people would expect that west can tell you if your branch in up-to-date (when tracking a branch).

I'm not 100% sure what you mean here but "up-to-date" is meaningless in optimistic version control[*] like git anyway because the remote branch moves all the time for obvious concurrency and productivity reasons. That's why it's called "tracking". Anyone missing this fact is in big version control trouble long before they even come anywhere near west. west is a git wrapper but it's not the kind of git wrapper meant to simplify the native git interface and concepts.

[*] https://thombergs.wordpress.com/2009/03/19/optimistic-vs-pessimistic-version-control/

> and if any of the west project are off == indication that SHA could be a local-only SHA)

Typically but not just.

tejlmand commented:
> Comparing to manifest-rev seems to be what this PR does (no time for a detailed review sorry).

Correct, because in this case the important information is the SHA, so that you know how to reproduce this particular build.
Using the meta-file together with `.config`, `zephyr.dts` and `CMakeCache.txt` will allow you to reproduce the build.

But you can only reproduce the build if you can ensure you have access to those SHAs.
That is why:
`-dirty` is important, as to know if the SHA uniquely identifies what was built.
`-extra` was extra stuff, besides the projects described in the manifest used for this build.
`-off` the SHA for this repo may not be reachable, even if you try to fetch the exact mentioned SHAs in the metafile.

So from a reproducibility point of view, the SHAs are required, and the extra knowledge are important to have in mind before considering if you will be able to reproduce this build.

The details described here https://github.com/zephyrproject-rtos/west/issues/548#issuecomment-938848138 doesn't apply in order to achieve the specific goal of the meta file, where the purpose is purely the possibility to recreate a specific build.

Note, for the particular `off` state, the SHA might be reachable but don't count on it.

> > and if any of the west project are off == indication that SHA could be a local-only SHA)
>
> Typically but not just.

correct, but that's not the point. 
The point is to make awareness that this SHA **may not** be available.
The fact that it can be available is just a bonus.

> Anyone missing this fact is in big version control trouble long before they even come anywhere near west. 

We are not all native english speakers, so the fact that I may not be precise in my description is not the same as not understanding how west and especially git works.

marc-hb commented:
> But you can only reproduce the build if you can ensure you have access to those SHAs. That is why:
> -dirty is important, as to know if the SHA uniquely identifies what was built.
> -extra was extra stuff, besides the projects described in the manifest used for this build.
> -off the SHA for this repo may not be reachable, even if you try to fetch the exact mentioned SHAs in the metafile.

Thanks for your patience, I think I finally get how the `-off` rationale here differs from  zephyrproject-rtos/west#548 - even when it ends up working exactly the same. I suspect part of my limited understanding earlier was because the first two items in your list are very different from the 3rd. The first two items are not about accessing the SHA listed in the lock file / meta file, they're about accessing the content local to and found only on the build system. They're all important of course.

> We are not all native english speakers, so the fact that I may not be precise in my description is not the same as not understanding how west and especially git works.

I'm afraid I'm lost even more. I just want to understand and assess the risk you mentioned. If west (or zephyr) makes no difference to how git beginners misunderstand git then that risk is off-topic.

marc-hb reviewed (commented):
This comment was left Pending by Github, not sure why...

marc-hb commented:
> Thanks for your patience, I think I finally get how the -off rationale here differs from zephyrproject-rtos/west#548 - even when it ends up working exactly the same [comparing with `manifest-rev`]

It ends up the same because the rationales are close:

1. The rationale for zephyrproject-rtos/west#548  was "Which SHA(s) am I building that make me different from everyone else?"
2. The rationale for `-off` here is: "Which SHA(s) did I build that everyone else _may_ not find?"

A simple difference between the 1. and 2. rationales is: `git checkout manifest-rev~N`

Because it's hard to make certain of 2., `-off` ends up implementing 1 instead and describes itself as "_maybe_ off the beaten path".

mbolivar-nordic commented on Kconfig.zephyr:
```diff
@@ -512,9 +512,35 @@ config BUILD_OUTPUT_META
+	    - dirty: one or more repositories are marked dirty
+	    - extra: extra Zephyr modules are manually included in the build
+	    - off:   the SHA of one or more west projects has diverged since the
+	             last `west update`.
```
> I think this help section has been trying very hard to avoid mentioning the term manifest-rev and provide some user-friendlier description instead, but that effort has failed.

I really agree with this

marc-hb replied:
> avoid mentioning the term manifest-rev and provide some user-friendlier description instead

Actually, the two are not mutually exclusive. A very short and semi-accurate description may not hurt as long as the handle to the complete, accurate and unambiguous definition is there too.

tejlmand replied:
> I think this help section has been trying very hard to avoid mentioning the term manifest-rev 

No that was not the reason.
The main reason was that this is already a pretty long Kconfig help text, and therefore I try to just have a brief description for each state.

I took @mbolivar-nordic proposal from here: https://github.com/zephyrproject-rtos/zephyr/pull/40085#discussion_r753616469 but left out the part inside the parenthesis.
Feel free to rewrite to use the `manifest-rev` directly if you prefer, but I felt the way it was split in @mbolivar-nordic proposal with text and extra parenthesis sentence was not needed for a brief help text.

marc-hb replied:
```
	    - off:   the SHA of one or more west projects are not what the
	               manifest defined when `west update` was run the last
                       time (`manifest-rev`).
```
Any sentence really as long as it includes the critical keyword `manifest-rev` that saves a longer description. This suggestion for instance omits the `west update proj1 proj2` possibility; it may miss other corner cases too.

tejlmand replied:
@marc-hb thanks.

Kconfig help updated to proposed text.
(and commit messages has detailed state description removed to avoid them getting out-of-sync :wink:  )

tejlmand commented:
> 2\. The rationale for `-off` here is: "Which SHA(s) did I build that everyone else _may_ not find?"

Yes for reproducibility.
But the `-off` information also serve as a warning.

For example: 
```
west:
  manifest: /projects/github/ncs/zephyr/west.yml
  # ^^^ This is the manifest file used
  projects:
  - path: /projects/github/ncs/zephyr
    revision: e0d2aea64437fa1fccab026289c2377c65977ad7
    # ^^^ This is the revision of the manifest repo
....
  - path: /projects/github/ncs/bootloader/mcuboot
    revision: 225b02468bf77806bdd5f503725256cb345970d2-off
    # ^^^ Warning: HEAD != manifest-rev, so the revision in use
    #              has not been checked out by `west update`
....
```

Tells the user that this build was not done as a pure update + build procedure:
```
$ git checkout -C<manifest-repo> <revision>
$ west update
$ west build <app>
```

so it also acts as kind of a warning on the purity of the build.

> `-off` ends up implementing 1 instead and describes itself as "maybe off the beaten path".

Not completely.
Because the `-off` is appended to the SHA.

So when you see: `225b02468bf77806bdd5f503725256cb345970d2-off`, then it is the SHA `225b02468bf77806bdd5f503725256cb345970d2` that is off compared to a reference point, which is `manifest-rev`.

Now, let's looks at `west list` default output ("{name:12} {path:28} {revision:40} {url}"), but include the SHA also:
```
$ west list --format="{name:12} {path:28} {revision:40} {sha:40} {url}"
manifest     nrf                          HEAD                                     N/A                                      N/A
zephyr       zephyr                       v2.6.99-ncs1-branch                      18e072b03270436fcd58312217fa8f362abb5ccd https://github.com/nrfconnect/sdk-zephyr
mcuboot      bootloader/mcuboot           800d4756702f07364f32538a6f0263c16be3eafc 800d4756702f07364f32538a6f0263c16be3eafc https://github.com/nrfconnect/sdk-mcuboot
nrfxlib      nrfxlib                      v1.8.0-rc2                               501d6b3a534d698090984043a4f583cbb846e2d1 https://github.com/nrfconnect/sdk-nrfxlib
...
```
Let's start examine mcuboot which is a SHA.

Let's checkout a different revision on `mcuboot`.
```
$ git -C ../bootloader/mcuboot checkout 225b02468bf77806bdd5f503725256cb345970d2
...
HEAD is now at 225b0246 merge: synchronized with mcutool/mcuboot 1a9c6d8
$ west list --format="{name:12} {path:28} {revision:40} {sha:40} {url}" mcuboot
mcuboot      bootloader/mcuboot           800d4756702f07364f32538a6f0263c16be3eafc 800d4756702f07364f32538a6f0263c16be3eafc https://github.com/nrfconnect/sdk-mcuboot
```
as you see, the revisions reported by `west list` are not changed because `west list` doesn't consider `HEAD` of the repo.

I think it would be wrong to add `-off` in this case, for example: `800d4756702f07364f32538a6f0263c16be3eafc-off`
Why would that be wrong ? Because we don't have a reference point.
It's not `800d4756702f07364f32538a6f0263c16be3eafc` that is off to something.
With `west`, our reference point is the manifest itself, or `manifest-rev`.
But the `manifest-rev` is still pointing to that SHA. 

And saying `off` compared to HEAD is also bad, cause HEAD is not really a reference point, so would be better to say it's just different. But remember `west list` doesn't include HEAD anywhere.

But in the `metafile` where we have built something, then we can say a SHA is off compared to the `manifest-rev` because the manifest-rev is controlled by west using the west manifest, and the manifest is our reference.

Let's also examine Zephyr, which is tracking a branch:
```
$ git -C ../zephyr checkout 58234d490c32811860ffc818d662730448c713a7
...
HEAD is now at 58234d490c samples: subsys: mcumgr: Remove deprecated <power/reboot.h> header
$ west list --format="{name:12} {path:28} {revision:40} {sha:40} {url}" zephyr
zephyr       zephyr                       v2.6.99-ncs1-branch                      18e072b03270436fcd58312217fa8f362abb5ccd https://github.com/nrfconnect/sdk-zephyr
```
again, the SHA listed by `west list` is reporting `manifest-rev` not HEAD.
Adding off here `v2.6.99-ncs1-branch-off` could end up confusing users, cause what would `off` mean in such a case ?

`west list` reports both the remote `url` and the `branch` so this could falsely lead users to believe local revision is different from tip of the remote branch. This is what I mean by saying: `there is a risk that people would expect that west can tell you if your branch in up-to-date`.

So marking something `off` in `west list` could easily confuse users in some cases.
Especially those not into the finer details of `git` and `west`. (@marc-hb I don't think you would be confused)

>  I just want to understand and assess the risk you mentioned.

Fair enough.
Maybe I misread your comment, but the way it was written regarding `git` and `west` and the links provided I felt you wanted to lecture, and not a real desire to understand my concerns of bringing this feature directly into `west list`.

I hope this additional comment answers your question (and doesnt't create new ones).

marc-hb commented:
> So marking something off in west list could easily confuse users in some cases.

I'm aware some people have mentioned implementing https://github.com/zephyrproject-rtos/west/issues/548 in `west list` (I have not) but I don't think anyone has suggested changing existing fields like what you described. I'm confused; I feel like you reviewing and objecting to `west list` proposals that don't exist.

> It's not 800d4756702f07364f32538a6f0263c16be3eafc that is off to something.

Indeed not, it's the `225b02468bf77806bdd5f503725256cb345970d2-off` HEAD that is off from `800d4756702f07364f32538a6f0263c16be3eafc` if I understand your long example(s) correctly.

> Adding off here v2.6.99-ncs1-branch-off could end up confusing users

I don't think anyone ever suggested that 548 made sense for branches either. It does not.

548 is really just "give me the complete `git status` like git does git submodules", it's literally the second sentence in the description at top. West does a number of things that git submodules do not; so 548 does not apply there. West does a number of things differently from git submodules, which is why the only simple conclusion from 548 was (IMHO) "just compare HEAD with manifest-rev and report that, end of story. Keep it simple and no one gets confused".

tejlmand commented:
> but I don't think anyone has suggested changing existing fields like what you described.

Well, you asked: `More generally speaking, could the non-Zephyr specific features being added here be later moved outside Zephyr and made available to all west users?` https://github.com/zephyrproject-rtos/zephyr/pull/40085#issuecomment-985206261

and this feature is appending the state `-off` to the `sha` (revision).
And then I repplied that yes, it can be done, and then followed up with some reasons as to why we should be careful doing so, https://github.com/zephyrproject-rtos/zephyr/pull/40085#issuecomment-988638900.

You didn't specify whether `-off` should only be applied in some cases, or exactly how it should be made available.
You said `non-Zephyr specific features` and `made available to all west users`.


> > Adding off here v2.6.99-ncs1-branch-off could end up confusing users
> 
> I don't think anyone ever suggested that 548 made sense for branches either. IMHO it does not.

See above regarding your initial question. ^^^
It should be clear from my description that I don't think it makes sense, but that was not your initial question so how was I to tell if that covered branches / PRs revisions ? It's clear now.

On top of that, there is also this proposal to add status for all, like this:
https://github.com/zephyrproject-rtos/west/issues/548#issuecomment-940110362

This PR $40085 only has one state, so if doing: `made available to all west users?` is not what you mean, then you should describe it clearer.

But if you want to discuss a solution to https://github.com/zephyrproject-rtos/west/issues/548 then please do it in that issue (or a new issue), not in this PR, cause this PR is not trying to solve  https://github.com/zephyrproject-rtos/west/issues/548

marc-hb commented:
> Well, you asked: More generally speaking, could the non-Zephyr specific features being added here be later moved outside Zephyr and made available to all west users?

Yes, and I mentioned 548 first before that (and then in almost every comment)

>  cause this PR is not trying to solve zephyrproject-rtos/west#548

I noticed it compares HEAD and manifest-rev, so I think it solves 548 better than it solves itself (no: "maybe" -off)

tejlmand commented:
> > Well, you asked: More generally speaking, could the non-Zephyr specific features being added here be later moved outside Zephyr and made available to all west users?
> 
> Yes, and I mentioned 548 first before that (and then in almost every comment)

You do, but 548 describes an issue and discusses possible solutions but doesn't conclude.
This PR presents a particular implementation for build meta info constructed at build time and uses HEAD sha with `-off` as postfix. I still fail to see how you expect this to be moved to `west` as common feature.

> 
> > cause this PR is not trying to solve [zephyrproject-rtos/west#548](https://github.com/zephyrproject-rtos/west/issues/548)
> 
> I noticed it compares HEAD and manifest-rev, so I think it solves 548 better than it solves itself (no: "maybe" -off).

It does, but how should this comparison be used and presented ?
Please open an issue here: https://github.com/zephyrproject-rtos/west/

Where you describe you would like this feature added to `west` and include:
- How should this be presented to the user.
- Which command should provide this info, new command or `west list` or ... ?
- Should west start to present a repo's HEAD.

I really fail to see how this feature should be moved into `west` and so far you have not provided any details of how you see this accomplished.

marc-hb commented:
> I really fail to see how this feature should be moved into west and so far you have not provided any details of how you see this accomplished.

I haven't because that is... the question I asked in the first place. The two conclusions of the discussion here are:
- The  logic implemented here fits 548 (even better than it fits itself IMHO)
- As you just described, the user interface is pretty far from what 548 would need. Maybe it's possible to create spurious meta files as a workaround in the mean time. Worth a try.

I'm sorry about the misunderstandings and that it went such a long and convoluted way but these two points answer my initial question in the end, thanks!

> Please open an issue here: https://github.com/zephyrproject-rtos/west/

From https://github.com/zephyrproject-rtos/west/issues/548#issuecomment-947720138

>  I will close this (548) then, but please feel free to reopen if you (or anyone else) feels they have a general solution backed up by experience and testing.

(.... which this meta file may provide some of)
